src/app/clusters/temperature-control-server/temperature-control-server.cpp

```diff
@@ -0,0 +1,230 @@
+/**
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ *
+ */
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/temperature-control-server/supported-temperature-levels-manager.h>
+#include <app/util/attribute-storage.h>
+#include <app/util/error-mapping.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TemperatureControl;
+using namespace chip::app::Clusters::TemperatureControl::Attributes;
+using namespace chip::DeviceLayer;
+using chip::Protocols::InteractionModel::Status;
+
+namespace {
+
+const uint8_t kMaxTemperatureLevelStringSize = 32;
+
+static SupportedTemperatureLevelsIteratorDelegate * sInstance = nullptr;
+
+class TemperatureControlAttrAccess : public AttributeAccessInterface
+{
+public:
+    // Register for the TemperatureControl cluster on all endpoints.
+    TemperatureControlAttrAccess() : AttributeAccessInterface(Optional<EndpointId>::Missing(), TemperatureControl::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+};
+
+} // namespace
+TemperatureControlAttrAccess gAttrAccess;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TemperatureControl {
+
+SupportedTemperatureLevelsIteratorDelegate * GetInstance()
+{
+    return sInstance;
+}
+
+void SetInstance(SupportedTemperatureLevelsIteratorDelegate * instance)
+{
+    sInstance = instance;
+}
+
+} // namespace TemperatureControl
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+CHIP_ERROR TemperatureControlAttrAccess::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    VerifyOrDie(aPath.mClusterId == TemperatureControl::Id);
+
+    if (TemperatureControl::Attributes::SupportedTemperatureLevels::Id == aPath.mAttributeId)
+    {
+        TemperatureControl::SupportedTemperatureLevelsIteratorDelegate * instance = TemperatureControl::GetInstance();
+        if (instance == nullptr)
+        {
+            aEncoder.EncodeEmptyList();
+            return CHIP_NO_ERROR;
+        }
+
+        instance->Reset(aPath.mEndpointId);
+
+        CHIP_ERROR err;
+        err = aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
```

Comment:
The fact that errors here are not being propagated out correct is why CI is failing.

Reply:
Fixed it. Thank you